File: Data Structures & Algorithms/buy-and-sell-crypto/submission-1.java
class Solution {
    public int maxProfit(int[] prices) {
        int l=0,r=0;
        int maxProfit = 0;

        for(int i=0;i<prices.length;i++){
            if(prices[l]>=prices[r]){
                l=r;
            }else{
                int profit = prices[r] - prices[l];
                maxProfit = Math.max(maxProfit, profit);
            }
            r++;
        }
        return maxProfit;
    }
}

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-1.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        dp = {}
        def dfs(i, canBuy):
            if i>= len(prices):
                return 0
            if(i, canBuy) in dp:
                return dp[(i,canBuy)]
            if canBuy:
                buy = dfs(i+1, not canBuy) - prices[i]
                cooldown = dfs(i+1, canBuy)
                dp[(i,canBuy)] = max(buy, cooldown)
            else:
                sell = dfs(i+2, not canBuy) + prices[i]
                cooldown = dfs(i+1, canBuy)
                dp[(i, canBuy)] = max(sell, cooldown)
            return dp[(i,canBuy)]
        return dfs(0, True)